### PR TITLE
Rough implementation of json to avro bytes conversion

### PR DESCRIPTION
**Pull request by eugene-cheverda**

### CODE REVIEW

File: example/src/main/scala/pulse/services/example/api/v1/status/StatusApi.scala [hunk 1]
```diff
@@ -1,16 +1,33 @@
...
+import pulse.services.example.avro.AvroUtils
+import pulse.services.example.avro.AvroUtils.jsonToAvroBytes
+import shapeless.{:+:, CNil}
+import java.lang._
 /**
   * Created by Andrew on 16.03.2017.
   */
 object StatusApi {
-  def statusApi() = status
+  def statusApi(): Endpoint[:+:[Status, :+:[String, CNil]]] = status :+: updateStatus
```
Comment (theocat) [inline example/src/main/scala/pulse/services/example/api/v1/status/StatusApi.scala]: creepy :)

File: example/src/main/scala/pulse/services/example/api/v1/status/StatusTaskApi.scala [hunk 2]
```diff
@@ -1,17 +1,28 @@
...
-  def statusApi() = status
+  def statusApi() = status :+: updateStatus
 
   def status: Endpoint[Status] =
     get("v1" :: "status") {
       _transformTask(Task.now(Ok(Status("fine"))))
     }
+
+  def updateStatus: Endpoint[String] = post("v1" :: "status" :: stringBody) {
+    s: String => {
```
Comment (theocat) [inline example/src/main/scala/pulse/services/example/api/v1/status/StatusTaskApi.scala]: The entire thing is supposed to be run inside a future to unblock the endpoint

File: example/src/main/scala/pulse/services/example/avro/AvroUtils.scala [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+import org.apache.avro.generic.{GenericData, GenericDatumReader, GenericDatumWriter, GenericRecord}
+import org.apache.avro.data.Json.ObjectReader
+import org.apache.avro.file.{DataFileReader, DataFileWriter, FileReader}
+import org.apache.avro.generic.GenericData.Record
+import org.apache.avro.io.{DecoderFactory, EncoderFactory}
+import org.apache.avro.specific.SpecificDatumWriter
+
+case object AvroUtils {
+
+  def jsonToAvroBytes(json: String, schema:String): Array[Byte] = {
```
Comment (theocat) [inline example/src/main/scala/pulse/services/example/avro/AvroUtils.scala]: it's very unsafe in terms of resource management. should be used with brackets

File: example/src/main/scala/pulse/services/example/api/v1/status/StatusTaskApi.scala (hunk 2 above)
Comment (eugene-cheverda) [inline example/src/main/scala/pulse/services/example/api/v1/status/StatusTaskApi.scala]: It is not completed yet, it's only a test code for json to avro serialization, once we have kafka producer implementation I'll wrap everything into Future as well

File: example/src/main/scala/pulse/services/example/avro/AvroUtils.scala (hunk 3 above)
Comment (eugene-cheverda) [inline example/src/main/scala/pulse/services/example/avro/AvroUtils.scala]: Fixed

File: example/src/main/scala/pulse/services/example/api/v1/status/StatusApi.scala (hunk 1 above)
Comment (eugene-cheverda) [inline example/src/main/scala/pulse/services/example/api/v1/status/StatusApi.scala]: Fixed

**Comment by eugene-cheverda** [comment]
Please review

File: example/src/main/scala/pulse/services/example/api/v1/status/StatusTaskApi.scala [hunk 4]
```diff
@@ -1,17 +1,36 @@
...
+      val bytes = AvroUtils.jsonToAvroBytes(s, "/Users/eugene/work/pulse/fgpservices/example/src/main/scala/pulse/services/example/status.avsc")
```
Comment (theocat) [inline example/src/main/scala/pulse/services/example/api/v1/status/StatusTaskApi.scala]: Please do not use anything hardcoded. Use a cmd parameter from scopt to provide a path

File: example/src/main/scala/pulse/services/example/api/v1/status/StatusTaskApi.scala [hunk 5]
```diff
@@ -1,17 +1,36 @@
...
+      bytes match {
+        case Right(b) => _transformTask(Task.now(Ok(new String(b, Charset.defaultCharset())))
+        case Left(ex) => {
+          print(s"Error during json to avro serialization: ${ex.getMessage}")
+          _transformTask(Task.now(InternalServerError(new Exception(ex.getMessage)))
```
Comment (theocat) [inline example/src/main/scala/pulse/services/example/api/v1/status/StatusTaskApi.scala]: Task.fail

File: example/src/main/scala/pulse/services/example/avro/AvroUtils.scala [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+import org.apache.avro.Schema
+import org.apache.avro.file.DataFileWriter
+import org.apache.avro.generic.{GenericDatumReader, GenericDatumWriter, GenericRecord}
+import org.apache.avro.io.DecoderFactory
+import pulse.services.example.extensions._
+
+case object AvroUtils {
+
+  def jsonToAvroBytes(json: String, schema:String) = {
+    val schemaSpec = loadSchema(schema)
```
Comment (theocat) [inline example/src/main/scala/pulse/services/example/avro/AvroUtils.scala]: what if this part throws?

File: example/src/main/scala/pulse/services/example/avro/AvroUtils.scala [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+
+import java.io.{ByteArrayInputStream, ByteArrayOutputStream, DataInputStream, File}
+
+import org.apache.avro.Schema
+import org.apache.avro.file.DataFileWriter
+import org.apache.avro.generic.{GenericDatumReader, GenericDatumWriter, GenericRecord}
+import org.apache.avro.io.DecoderFactory
+import pulse.services.example.extensions._
+
+case object AvroUtils {
```
Comment (theocat) [inline example/src/main/scala/pulse/services/example/avro/AvroUtils.scala]: why case object?

File: example/src/main/scala/pulse/services/example/extensions/package.scala [hunk 8]
```diff
@@ -0,0 +1,23 @@
+package pulse.services.example
+
+import java.io.Closeable
+
+package object extensions {
+
+    def bracket[A <: Closeable, B](obj: A)(body: A => B) = {
```
Comment (theocat) [inline example/src/main/scala/pulse/services/example/extensions/package.scala]: I would strongly suggest you to use implicits instread. Let it be like:
```
def bracket[A,B](obj: A)(body: A => B)(implicit C: Closeable[A]) = {
     try {
        Right(body(obj))
     }
    catch {
        case e => Left(e)
    }
    finally {
         C.close(obj)
    }
}

```

File: example/src/main/scala/pulse/services/example/avro/AvroUtils.scala (hunk 7 above)
Comment (eugene-cheverda) [inline example/src/main/scala/pulse/services/example/avro/AvroUtils.scala]: Addressed

File: example/src/main/scala/pulse/services/example/api/v1/status/StatusTaskApi.scala (hunk 4 above)
Comment (eugene-cheverda) [inline example/src/main/scala/pulse/services/example/api/v1/status/StatusTaskApi.scala]: Addressed

File: example/src/main/scala/pulse/services/example/avro/AvroUtils.scala (hunk 6 above)
Comment (eugene-cheverda) [inline example/src/main/scala/pulse/services/example/avro/AvroUtils.scala]: Addressed

File: example/src/main/scala/pulse/services/example/extensions/package.scala (hunk 8 above)
Comment (eugene-cheverda) [inline example/src/main/scala/pulse/services/example/extensions/package.scala]: Addressed

File: example/src/main/scala/pulse/services/example/api/v1/status/StatusTaskApi.scala (hunk 5 above)
Comment (eugene-cheverda) [inline example/src/main/scala/pulse/services/example/api/v1/status/StatusTaskApi.scala]: Addressed

**Comment by eugene-cheverda** [comment]
Please review

**Pull request merged**